### ISSUE TITLE
OCaml 5.0 & unhandled exceptions

### ISSUE THREAD

**Issue by dbuenzli** [ocaml/ocaml#11074]
While dabbing a bit into the future offering of OCaml 5, I got a bit confused on how unhandled exceptions are supposed to be handled among all the various threads of execution that became available.

One thing that surprised me is `Domain.join` raising the exception that may have been thrown by its `Domain.spawn` function. I thought the idea of domains was rather to be rare (in the sense one per core) "spawn and forget" like entities. Was the idea that domains would always periodically `join` their `spawn`s ? But then I can only `join` one domain and it blocks !

So I would rather have expected each of them to have their own `Printexc.set_uncaught_exception_handler` initialized from the one set in the main domain.  Unless I misunderstood something what we are looking at here is uncaught exceptions in domains going unnoticed (except for the main one) which I have to say is not an enticing perspective for long running programs.

It also seems that the `Thread` module now got its own `Thread.set_uncaught_exception_handler`. That's certainly a good addition since previously threads would simply ignore `Printexc`'s one, but from the docs I couldn't understand if you have to set that per thread or if it uses the one you specified once for newly created threads.

Is there an actual design that I failed to see here ? From an end-user perspective I find all that quite confusing, are we going to need to set multiples places with uncaught exception handlers for long running programs ? 

If you asked me it would be time to take the opportunity to define that `Exn` module [we talked about during quite some time](https://github.com/ocaml/ocaml/pull/2137) before it got closed without explanations and (if possible) centralize the reporting of uncaught exception with via a single handler that indicates in which context the uncaught exception occurred (bonus point if the context has room for user cases).

**Comment by kayceesrk** [comment]
@dbuenzli 

(On my phone; please excuse the lack of code formatting)

I view Domain.join to be analogous to awaiting on a promise. The normal or the exceptional result from the promise should be given to those who wait on the promise. 

The general assumption (not written down, but should be if true) is that Domain.spawn is not a fire-and-forget mechanism. We expect all spawned domains to be joined. I can’t think of a crash if the main domain quits before child domains, but @Engil, @sadiq and @ctk21 have thought about it more than I have. 

There are indeed unexpected behaviours if the main domain quits without joining on all the children. For example, the predefined buffers in the Format library from the child domains will not be flushed. 

I was not aware of the unexpected exception handler in the Thread module. Thanks for bringing this to my attention. I’m uncomfortable with only having a callback for handling exceptions raised by domains, especially if the guarantee that every spawned domain must be joined. I’m not wedded to the idea that Domain.join raises the same exception as the child domain. Perhaps returning a ‘a result value would work?

As a proposal, if we expect all spawned domains to be joined (makes some termination behaviour in the runtime easier I would think) would you be happy to live with just the current behaviour, or returning a ‘a result on Domain.join? Again, I dislike having callbacks for exceptional behaviour as one needs to reason non-locally about program behaviour.

**Comment by dbuenzli** [comment]
> I view Domain.join to be analogous to awaiting on a promise. The normal or the exceptional result from the promise should be given to those who wait on the promise.

Nothing says that awaiting on a promise should raise an exception if the latter does while determining it. In fact I would rather argue that it is a bad idea, as `lwt` messy design w.r.t. exceptions and cancellation has shown us.

The problem is that in general there is no mecanism to distinguish whether an exception is a spurious one or one that is raised by design because the client wants to work with exceptions. 

I think it is important to be able to make that distinction with parallel and concurrent threads of execution (i.e. in Promises/Fibers/Domains/Threads) where, in contrast to simple functions calls,  lots of "effectful stuff" may be happening for an arbitrary long periods of time. Handling spurious exceptions occuring in these activities should not be on the thread of execution that joins them.

The only way I see to make that distinction is to let Promises/Fibers/Domains determine *values*, trap uncaught exceptions (and thus let `Domain.join : 'a t -> 'a option`) and for people who want to work with exception by design let their domain spawn functions return an `('a, exn) Either.t` value.

> The general assumption (not written down, but should be if true) is that Domain.spawn is not a fire-and-forget mechanism.

Then the current interface seems unpractical. Unless I misunderstood something I can't join domains in parallel. Moreover calling `Domain.join` means blocking the calling thread. Blocking the main thread is forbidden in certain settings (e.g. in UIs on `macos` or `ios`).

To take an example, suppose I spawn a domain that is going to handle ad infinitum all my network with one peer and another domain with another peer. How can I: 

1. Detect if a spurious exception killed either of these. 
2. Do so without blocking the main thread.

I guess at least for 2. you could spawn a domain that spawns the domains. But at that point aren't we just working around bad design ? I think at least a `Domain.poll` for checking the termination status of a domain should be added.

> Again, I dislike having callbacks for exceptional behaviour as one needs to reason non-locally about program behaviour.

Not sure what the objection is here. Such callbacks already need to exist anyways. They are not meant to be used they are here as a last resort mecanism so that no errors go unnoticed. 

> or returning a ‘a result on Domain.join?

As mentioned above having something explicit to distinguish exceptional termination is certainly a good idea. 

But in doing so I would still strongly argue for not being able to determine what exceptional terminations are (i.e. `'a result` could simply be `type 'a result = 'a option`) and simply give those to a trap.

Otherwise, if you let people access the exception at the `Domain.join` boundary you:

1. Push again the uncaught exception problem on the joining client (which may entirely ignore it and thus let spurious errors go unnoticed)
2. No longer allow to have a clear distinction between spurious and by design exceptions.

**Comment by kayceesrk** [comment]
>  In fact I would rather argue that it is a bad idea, as lwt messy design w.r.t. exceptions and cancellation has shown us.

Lwt's problem with exceptions is that the stack discipline is different (callback-oriented programming) which makes it incompatible with built-in exceptions in OCaml. I don't see why lwt is relevant here since domains don't suffer this problem. Perhaps you are referring to some other specific issue with lwt cancellation semantics. 

> To take an example, suppose I spawn a domain that is going to handle ad infinitum all my network with one peer and another domain with another peer. How can I:
>
> Detect if a spurious exception killed either of these.
> Do so without blocking the main thread.

Use shared mutable state? For instance, `type status = Running | Done | Failed of exn` and have a flag each for the two domains of type `status Atomic.t`? Let the child domains update their status on termination? 

>  I think at least a Domain.poll for checking the termination status of a domain should be added.

This seems reasonable. 

## Separation of policies and mechanisms

Let me know if I am wrong but I feel like you have a particular policy for handling spurious exceptions, which is one of many, and are proposing a mechanism for it. I don't see why such a mechanism should be provided by the `Domain` module when it could be implemented fairly easily using other mechanisms provided in the standard library. 

I guess one way to think about this is to separate out the "primitive" features that must be provided by the `Domain` module from the non-primitive ones which may be implemented on top. For OCaml 5.0, we don't know how users will use the new features. So our aim was to not include many non-primitive ones in the first release, and then, based on need include more non-primitive utility functions in the module. 

You may argue that `Domain.join` is not a primitive feature and is subscribing to a particular policy. Indeed, one could implement that using `Mutex` and `Condition` (which is not how it is implemented, and I see no reason why not).

## Uncaught exception handlers

Regarding uncaught exception handlers, if we propose another uncaught exception handler in the `Domain` module, I'm unclear when the uncaught exception handlers of `Printexc`, `Thread` and the `Domain` would apply. For example, in my sequential program, if I install uncaught exception handlers using all the three modules, would all of them be invoked, or just one (is there a priority between the three modules, or is it the order in which these callbacks were installed). You can ask similar questions about installing uncaught exception handlers for combinations of domains and systhreads. I am reluctant to pursue this option until this design becomes clear. 

@Engil, since you'd implemented the uncaught exception handlers for threads, perhaps you can shed light on the interaction between `Printexc`'s and `Thread`'s one. What is the intended semantics?

**Comment by dbuenzli** [comment]
> I don't see why lwt is relevant here since domains don't suffer this problem.

Well you mentioned promises as an inspiration for the semantics of `Domain.join`, so here it goes. 

Combinators like `pick` are interested in getting *values* not propagating the fact that a promise exceptionally terminated. Since `lwt` does not distinguish between the two this renders the combinators algebraically useless and even funnily absurd, given that cancellation was implemented using an exception.

Hence my stance on having a proper distinction between value and exceptional termination on entities that deal with long program activities.

> Use shared mutable state? For instance, `type status = Running | Done | Failed of exn` and have a flag each for the two domains of type `status Atomic.t`? Let the child domains update their status on termination?

Indeed that's possible, but from someone who wanted to avoid non-local reasoning, that's a rather surprising suggestion :–) I would rather suggest to give the programmer proper tools.

> Let me know if I am wrong but I feel like you have a particular policy for handling spurious exceptions, which is one of many, and are proposing a mechanism for it.

I don't think I have anything particular here. I'm just suggesting to extend the mecanisms that *already* exist so that they can cope with the new stuff that is being introduced. 

In particular I see it as a design flaw that uncaught exceptions in program activities (threads of executions) can go unnoticed for an arbitrary amount of time (namely before `Domain.join` is called), possibly before a crash. This has never been the case so far, I don't think it would be a good idea if it would become.

> I guess one way to think about this is to separate out the "primitive" features that must be provided by the `Domain` module from the non-primitive ones which may be implemented on top.

Some disciplines still need to be enforced at the API level if you want to be able to enforce it on a distributed eco-system. This often seen "We don't know how users will use the new features" trope is just a way to shy away from proper design which will only lead again to eco-system fragmentation and bad API usability.

> Regarding uncaught exception handlers, if we propose another uncaught exception handler in the `Domain` module, I'm unclear when the uncaught exception handlers of `Printexc`, `Thread` and the `Domain` would apply.

As I mentioned in my first message, the last thing I want is a proliferation of exception handlers. As far as I'm concerned the new additions in the `Thread` module are not a good idea. 

I don't know how feasible it is but I advocate for a *single* trap to be used by *any* program activity whenever it sees it fit. That is something like:

```ocaml
type Exn.context = Domain.id * Thread.id * string
val Exn.set_trap : (Exn.context -> exn -> raw_backtrace -> unit) -> unit
val Exn.trap : unit -> Exn.context -> exn -> raw_backtrace -> unit 
```

**Comment by kayceesrk** [comment]
Currently, the situation is that

| Thread | Uncaught exception handler |
|----|----|
| Main (1st) thread of the program | `Printexc` module's uncaught exception handler |
| Threads created by the `Thread` module | `Thread` module's uncaught exception handler | 
| Threads created by `Domain.spawn` | Uncaught exceptions are captured in `Domain.t` and can be extracted by `Domain.join` | 

This isn't great. I've been discussing with @Engil on a different channel and we feel that unifying the various uncaught exception handlers would be a good idea. 

The API you propose seems like a good place to start. One thing I'm not clear about is how we can provide `Thread.id` when the `Thread` module isn't linked.

**Comment by abbysmal** [comment]
> The API you propose seems like a good place to start. One thing I'm not clear about is how we can provide Thread.id when the Thread module isn't linked.

I think having a generalist trapping mechanism is indeed a nice option.
It would seem having a `Domain.id * Thread.t * string` tuple is not quite a good reflection of the underlying situation: If `systhreads` is not linked in, they would be no `Thread.t` anyway. (so this field should be optional)

My initial take would be that returning an integer for the `Thread.t` identifier (using `Thread.id`) is probably fine, because the user can likely map out whatever happened from there.
And assuming we return a `Domain.id`, this would be effectively the same as for the Domain side of things. (since both the integer out of `Thread.id` and the private integer of `Domain.id` can not be used to map back to a proper handle.)

Maybe `Thread` could benefit of mirroring the way `Domain` handles and identifiers work ?
By having a `private int` `Thread.id` type, instead of the current situation where we have a `Thread.id` function returning an integer?
This would however bring back the problem where having a `Thread.id` type would be an issue since `systhreads` is not always linked in.
We would ultimately need to expose an integer. So this is probably not a very useful observation.

**Comment by dbuenzli** [comment]
> Domain.id * Thread.t * string

I don't think people should trip too much on that signature which I proposed without knowing the stdlib dependency dag on trunk (and assuming there was nowadays always a thread id available for reporting). 

`type context = Domain.id * string` or even `type context = string` is just as fine. Any entity can serialize its identity/context to a string before using the trap. 

I think machine readability is of little interest here, this is a last resort for when things go wrong so that unexpected errors don't go unnoticed. In a good program in good conditions this should not happen.

**Comment by favonia** [comment]
Zooming out, I would like to echo @dbuenzli's opinion (in [another thread](https://github.com/ocaml/ocaml/pull/11423#issuecomment-1188804926)) that new entry points for OCaml 5.0 should be marked experimental. In particular, I think the `Domain` module could definitely benefit from more thinking. The module has recently received and will probably receive many improvements, but I feel we will not have reflected upon those changes enough when OCaml 5.0 is out. As a user, I will be much happier if there's still a chance to continue refining the `Domain` module in the following minor revisions (5.1, 5.2, ...) even if I have to rewrite my own libraries.

**Comment by xavierleroy** [comment]
> new entry points for OCaml 5.0 should be marked experimental. In particular, I think the Domain module could definitely benefit from more thinking

Agreed.  Perhaps we can make this clear(er) in the documentation and the release notes.  (@Octachron, what do you think?)  Still, the more we can improve the new APIs before OCaml 5.0 is released, the better !

**Comment by Octachron** [comment]
I think it is certainly worth it to make it clearer in the documentation that both modules are experimental. We could also possibly add an alert? Similarly, for the release note, I think it could be good idea to have a high-level section that describes why 5.0 is an experimental release (in term of architectures supported, effect system, and domain library at the very least).

**Comment by gasche** [comment]
Yes, we could have an `experimental` alert on Domain, Effects and Runtime_events.

**Comment by xavierleroy** [comment]
> we could have an experimental alert on Domain, Effects and Runtime_events.

Assuming the alert triggers by default, I find this unnecessarily harsh on our users.  I hope more of them are interested in trying the new features than in suing us because the interfaces change between 5.0 and 5.1.

**Comment by gasche** [comment]
Indeed, the idea would be to not enable it by default. I think the point of having the alert is that it's a clear, non-ambiguous way to mark stability non-guarantees. (We can "mark a function/module experimental" by adding the annotation, and "unmark it" by removing the annotation, each change is naturally a PR with an attached discussion. It's not a matter of communicating things in release notes or here or there, which we easily get wrong.)

**Comment by dbuenzli** [comment]
> to mark stability non-guarantees.

That wording made me think that `unstable` may be a better word than `experimental` for the alert.

Other than that +1 for having an alert but not enabled by default (which I think is also a useful tool for APIs in the larger eco-system). This makes it easy to implement policies like no use of unstable features in a project without bothering too much those projects that want to use them.

**Comment by kayceesrk** [comment]
We should keep this issue open. The fixes to these issues will be breaking changes and hence I sense a general hesitation to start working on a solution.

**Comment by gadmm** [comment]
Having had similar thoughts in my recent work on Domain.spawn:

> Perhaps returning a ‘a result value would work?

Having `Domain.join : 'a Domain.t -> ('a,exn) Result.t` is more useful, since places where you can legitimately catch all exceptions and wrap them in a result are rare, and the domain/thread closures are one of them.

In addition, some kind of custom handler makes sense because doing it library-side is not equivalent.

**Comment by dbuenzli** [comment]
> Having `Domain.join : 'a Domain.t -> ('a,exn) Result.t` is more useful,

You may want to reread [this message](https://github.com/ocaml/ocaml/issues/11074#issuecomment-1098914651).

**Comment by gadmm** [comment]
This is orthogonal to the question of having uncaught exception handlers. The user cannot mimick the result-returning behaviour because the catch-all handler would not catch all exceptions, for current implementation reasons.

(I am not convinced with `Domain.join : 'a Domain.t -> 'a option`, but it is only a suggestion at this point and not what I wanted to discuss in my message.)

While we are at it, can you please clarify what you have in mind regarding the distinction "spurious" vs. "by-design" exception, as I found that this might help clarify your messages? For instance, if an exception denotes the cancellation of the thread or domain, is it "spurious" or "by-design" (regardless of how one would implement a cancellation mechanism), and do you have other examples ?

Lastly, it is good to look at other languages. In Rust, [`join`](https://doc.rust-lang.org/std/thread/struct.JoinHandle.html#method.join) returns a `Result` with the payload of the panic as the error argument. There is also an `is_finished` predicate to implement a non-blocking join. Lastly, there is also a "panic hook", but it runs when a panic is raised, not when it becomes uncaught (e.g. at thread boundary).

**Comment by dbuenzli** [comment]
> While we are at it, can you please clarify what you have in mind regarding the distinction "spurious" vs. "by-design" exception, as I found that this might help clarify your messages?

Spurious means an unexpected and unhandled exception, for example an `Invalid_argument`, an `Assert_failure` or an exception you forgot to handle like a `Sys_error` you didn't think about.

By-design is when you use exceptions in your API design e.g. using the `Not_found` exception to denote the absence of a value. These exceptions should never make it to the uncaught exception handler you are expected either to not trigger them or to handle them for your programs to be correct; if only at the toplevel of your program.

> Lastly, it is good to look at other languages.

Before looking at what other languages do, let's try to figure out a good and ergonomic model for programmers.

Personally the model I'd like to work with is:

1. Tasks (domains/threads/fibers) are entities executing functions returning values (`unit -> 'a`).
2. If an exception tries to escape the function of a task, it is considered to be an unexpected exception, it gets trapped by a single global exception trap and the task returns no value, `None`.
3. If the task supports a notion of cancellation and gets canceled, the task gets aborted in some way an returns no value, `None`.
4. When tasks are joined they return the value computed by the task, that is `Some v` if the task function completed with a return value or `None` if there was an unexpected exception or a cancellation.

This means that as a user of tasks I have only two things to think about: whether the task completed with a value or whether it did not. 

This means that as a task function designer I need to be disciplined with my use of exceptions. Task functions should be self-contained as far as exception handling is concerned, like one does at the toplevel of the main program. The person who likes to use exception by-design can still use them by having a task function with the type `unit -> ('a, exn) Either.t`. This makes it clear to the task user that the task function is using exceptions by-design.

In contrast if you start propagating unexpected exceptions in 4. either by having `join` returning a result with the exception or by having the `join` reraise the unexpected exception in the joiner then: 

a) You offload the problem of what to do with them on tasks users. That's not a question I constantly want to have to answer and deal with when I'm using and coordinating tasks. I prefer to be able to see them as well delineated computation units that either produce a value or none.

b) You can't be sure whether the exception you are getting was unexpected or by-design by the person who designed the task function. The discipline becomes less clear.

c) In the particular case of domains you will not be informed about unexpected exception before you join the domain. That's a very bad idea: serious error conditions may go unnoticed, e.g. before a final catastrophic crash. You really want these things to show up asap in your logs – that is be trapped.
